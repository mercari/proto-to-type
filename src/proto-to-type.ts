import protobuf from 'protobufjs';
import { fetchProtoAll, FetcherOptions } from './fetch-proto';
import { createContext, GeneratorContextOptions } from './context';
import { resolveReflections } from './resolvers';
import { serialize, SerializerOptions } from './serialize';

export interface GeneratorOptions extends FetcherOptions, SerializerOptions, GeneratorContextOptions {}

export const ProtoToType = (options: GeneratorOptions) => {
  const generate = async (rootName: string, location: string, entrypoints: string[]) => {
    // step 1: fetch all related proto and accumulate reflections to the root
    const root = new protobuf.Root();
    const mainPackage = await fetchProtoAll(root, location, options);

    // step 2: prepare context
    const ctx = createContext(root, mainPackage, options);

    // step 3: look up entrypoints' reflection objects
    const entryReflections = entrypoints.map((entrypoint) => {
      const name = `${mainPackage}.${entrypoint}`;
      const ref = ctx.root.lookup(name);
      // eslint-disable-next-line no-console
      if (!ref) console.warn(`no resolved reflection found for ${name}`);
      return ref;
    });

    // step 4: resolve reflection objects recursively
    resolveReflections(ctx, entryReflections);

    // step 5: serialize context
    const content = serialize(ctx, [`declare namespace ${rootName} {`, '}'], options);

    Object.entries(ctx.unresolvedTypes).forEach(([name, count]) => {
      // eslint-disable-next-line no-console
      console.log(`'${name}' was not resolved correctly ${count} times.`);
    });

    return content;
  };

  return generate;
};
