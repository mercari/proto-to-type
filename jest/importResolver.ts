import { ImportResolver } from '../src';
import path from 'path';

// https://github.com/googleapis/googleapis/blob/50af0530730348f1e3697bf3c70261f7daaf2981/google/api/metric.proto
const GOOGLE_APIS_REPO = 'googleapis/googleapis';
const GOOGLE_APIS_COMMIT = '50af0530730348f1e3697bf3c70261f7daaf2981';
const METRIC = 'google/api/metric.proto';
const LABEL = 'google/api/label.proto';
const LAUNCH_STAGE = 'google/api/launch_stage.proto';

// https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/duration.proto
const GOOGLE_PROTOBUF_REPO = 'protocolbuffers/protobuf';
const GOOGLE_PROTOBUF_COMMIT = '176f7db11d8242b36a3ea6abb1cc436fca5bf75d';
const DURATION = 'google/protobuf/duration.proto';

export const protoFiles = { METRIC, LABEL, LAUNCH_STAGE, DURATION };

const fixtures = ['sample.proto', 'no-package.proto'];

export const importResolverForTesting: ImportResolver = (target) => {
  if (fixtures.includes(target)) {
    return ['read-file', path.resolve(__dirname, '../src/__fixtures__/', target)];
  } else if (target.startsWith('google/api/')) {
    return ['github-contents', GOOGLE_APIS_REPO, target, GOOGLE_APIS_COMMIT];
  } else if (target.startsWith('google/protobuf/')) {
    return ['github-contents', GOOGLE_PROTOBUF_REPO, `src/${target}`, GOOGLE_PROTOBUF_COMMIT];
  }
  return null;
};
