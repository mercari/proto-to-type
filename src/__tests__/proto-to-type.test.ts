import { ProtoToType, GeneratorOptions } from '../proto-to-type';
import { importResolverForTesting as resolveImport } from '../../jest/importResolver';
import wrap from 'jest-snapshot-serializer-raw';

const consoleLog = jest.spyOn(console, 'log');
const consoleWarn = jest.spyOn(console, 'warn');

beforeEach(() => {
  jest.clearAllMocks();
});

describe('ProtoToType', () => {
  it('generates type definition correctly', async () => {
    const generate = ProtoToType({
      resolveImport,
      predefinedTypes: { 'predefined.type': '100', string: 'string', int32: 'number', int64: 'number' },
    });
    const result = await generate('Sample', 'sample.proto', ['SampleService']);

    expect(consoleLog.mock.calls.length).toBe(0);
    expect(consoleWarn.mock.calls.length).toBe(0);
    expect(wrap(result)).toMatchSnapshot();
  });

  it('other formats', async () => {
    const args: Parameters<typeof generate> = ['Sample', 'sample.proto', ['SampleService']];
    const common: GeneratorOptions = {
      resolveImport,
      predefinedTypes: { 'predefined.type': '100', string: 'string', int32: 'number', int64: 'number' },
    };

    let generate = ProtoToType({ ...common, enumFormat: 'num' });
    expect(wrap(await generate(...args))).toMatchSnapshot('enumFormat: num');

    generate = ProtoToType({ ...common, enumFormat: 'mixed' });
    expect(wrap(await generate(...args))).toMatchSnapshot('enumFormat: mixed');

    generate = ProtoToType({ ...common, strictOneof: true });
    expect(wrap(await generate(...args))).toMatchSnapshot('strictOneof: true');

    generate = ProtoToType({ ...common, strictOneof: 'hybrid' });
    expect(wrap(await generate(...args))).toMatchSnapshot('strictOneof: hybrid');
  });

  it('warns if entrypoint wrongs', async () => {
    const generate = ProtoToType({
      resolveImport,
      predefinedTypes: { 'predefined.type': '100', string: 'string', int32: 'number', int64: 'number' },
    });
    const result = await generate('Sample', 'sample.proto', ['NeverService']);

    expect(consoleLog.mock.calls.length).toBe(0);
    expect(consoleWarn.mock.calls.length).toBe(1);
    expect(wrap(result)).toMatchSnapshot();
  });

  it('logs if unresolved type found', async () => {
    const generate = ProtoToType({
      resolveImport,
      predefinedTypes: { 'predefined.type': '100' },
    });
    const result = await generate('Sample', 'sample.proto', ['SampleService']);

    expect(consoleLog.mock.calls.length).toBe(3);
    expect(consoleWarn.mock.calls.length).toBe(0);
    expect(wrap(result)).toMatchSnapshot();
  });
});
