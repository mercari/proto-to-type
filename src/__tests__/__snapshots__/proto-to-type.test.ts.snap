// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtoToType generates type definition correctly 1`] = `
declare namespace Sample {
  namespace SampleService {
    namespace GetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/get";
      interface Response {
        data?: Messages.SampleData;
      }
      namespace Response {}
      interface Request {
        id?: string;
      }
      namespace Request {}
    }
    namespace SetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/set";
      interface Response {
        status?: Enums.SampleStatus;
      }
      namespace Response {}
      interface Request {
        id?: string;
        data?: Messages.SampleData;
      }
      namespace Request {}
    }
  }
  namespace Messages {
    interface SampleData {
      size?: number;
      id?: string;
      content?: string;
      status?: Enums.SampleStatus;
      createdAt?: Externals.google.protobuf.Timestamp;
      updatedAt?: Externals.google.protobuf.Timestamp[];
      extra?: 100;
      disabledAt?: Externals.google.protobuf.Timestamp;
      deletedAt?: Externals.google.protobuf.Timestamp;
      statusMap?: Record<string, Enums.SampleStatus>;
    }
    namespace SampleData {
      type sampleOneof =
        | { disabledAt: Externals.google.protobuf.Timestamp; deletedAt?: never }
        | { disabledAt?: never; deletedAt: Externals.google.protobuf.Timestamp }
        | { disabledAt?: never; deletedAt?: never };
    }
  }
  namespace Enums {
    type SampleStatus =
      | SampleStatus.UNKNOWN
      | SampleStatus.ACTIVE
      | SampleStatus.INACTIVE;
    namespace SampleStatus {
      type UNKNOWN = "UNKNOWN";
      namespace UNKNOWN {
        const str: "UNKNOWN";
        const num: 0;
      }
      type ACTIVE = "ACTIVE";
      namespace ACTIVE {
        const str: "ACTIVE";
        const num: 1;
      }
      type INACTIVE = "INACTIVE";
      namespace INACTIVE {
        const str: "INACTIVE";
        const num: 2;
      }
    }
  }
  namespace Externals {
    namespace google {
      namespace protobuf {
        interface Timestamp {
          seconds?: number;
          nanos?: number;
        }
        namespace Timestamp {}
      }
    }
  }
}

`;

exports[`ProtoToType logs if unresolved type found 1`] = `
declare namespace Sample {
  namespace SampleService {
    namespace GetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/get";
      interface Response {
        data?: Messages.SampleData;
      }
      namespace Response {}
      interface Request {
        id?: string;
      }
      namespace Request {}
    }
    namespace SetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/set";
      interface Response {
        status?: Enums.SampleStatus;
      }
      namespace Response {}
      interface Request {
        id?: string;
        data?: Messages.SampleData;
      }
      namespace Request {}
    }
  }
  namespace Messages {
    interface SampleData {
      size?: int32;
      id?: string;
      content?: string;
      status?: Enums.SampleStatus;
      createdAt?: Externals.google.protobuf.Timestamp;
      updatedAt?: Externals.google.protobuf.Timestamp[];
      extra?: 100;
      disabledAt?: Externals.google.protobuf.Timestamp;
      deletedAt?: Externals.google.protobuf.Timestamp;
      statusMap?: Record<string, Enums.SampleStatus>;
    }
    namespace SampleData {
      type sampleOneof =
        | { disabledAt: Externals.google.protobuf.Timestamp; deletedAt?: never }
        | { disabledAt?: never; deletedAt: Externals.google.protobuf.Timestamp }
        | { disabledAt?: never; deletedAt?: never };
    }
  }
  namespace Enums {
    type SampleStatus =
      | SampleStatus.UNKNOWN
      | SampleStatus.ACTIVE
      | SampleStatus.INACTIVE;
    namespace SampleStatus {
      type UNKNOWN = "UNKNOWN";
      namespace UNKNOWN {
        const str: "UNKNOWN";
        const num: 0;
      }
      type ACTIVE = "ACTIVE";
      namespace ACTIVE {
        const str: "ACTIVE";
        const num: 1;
      }
      type INACTIVE = "INACTIVE";
      namespace INACTIVE {
        const str: "INACTIVE";
        const num: 2;
      }
    }
  }
  namespace Externals {
    namespace google {
      namespace protobuf {
        interface Timestamp {
          seconds?: int64;
          nanos?: int32;
        }
        namespace Timestamp {}
      }
    }
  }
}

`;

exports[`ProtoToType other formats: enumFormat: mixed 1`] = `
declare namespace Sample {
  namespace SampleService {
    namespace GetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/get";
      interface Response {
        data?: Messages.SampleData;
      }
      namespace Response {}
      interface Request {
        id?: string;
      }
      namespace Request {}
    }
    namespace SetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/set";
      interface Response {
        status?: Enums.SampleStatus;
      }
      namespace Response {}
      interface Request {
        id?: string;
        data?: Messages.SampleData;
      }
      namespace Request {}
    }
  }
  namespace Messages {
    interface SampleData {
      size?: number;
      id?: string;
      content?: string;
      status?: Enums.SampleStatus;
      createdAt?: Externals.google.protobuf.Timestamp;
      updatedAt?: Externals.google.protobuf.Timestamp[];
      extra?: 100;
      disabledAt?: Externals.google.protobuf.Timestamp;
      deletedAt?: Externals.google.protobuf.Timestamp;
      statusMap?: Record<string, Enums.SampleStatus>;
    }
    namespace SampleData {
      type sampleOneof =
        | { disabledAt: Externals.google.protobuf.Timestamp; deletedAt?: never }
        | { disabledAt?: never; deletedAt: Externals.google.protobuf.Timestamp }
        | { disabledAt?: never; deletedAt?: never };
    }
  }
  namespace Enums {
    type SampleStatus =
      | SampleStatus.UNKNOWN
      | SampleStatus.ACTIVE
      | SampleStatus.INACTIVE;
    namespace SampleStatus {
      type _num = Extract<SampleStatus, number>;
      type _str = Extract<SampleStatus, string>;
      type UNKNOWN = "UNKNOWN" | 0;
      namespace UNKNOWN {
        const str: "UNKNOWN";
        const num: 0;
      }
      type ACTIVE = "ACTIVE" | 1;
      namespace ACTIVE {
        const str: "ACTIVE";
        const num: 1;
      }
      type INACTIVE = "INACTIVE" | 2;
      namespace INACTIVE {
        const str: "INACTIVE";
        const num: 2;
      }
    }
  }
  namespace Externals {
    namespace google {
      namespace protobuf {
        interface Timestamp {
          seconds?: number;
          nanos?: number;
        }
        namespace Timestamp {}
      }
    }
  }
}

`;

exports[`ProtoToType other formats: enumFormat: num 1`] = `
declare namespace Sample {
  namespace SampleService {
    namespace GetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/get";
      interface Response {
        data?: Messages.SampleData;
      }
      namespace Response {}
      interface Request {
        id?: string;
      }
      namespace Request {}
    }
    namespace SetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/set";
      interface Response {
        status?: Enums.SampleStatus;
      }
      namespace Response {}
      interface Request {
        id?: string;
        data?: Messages.SampleData;
      }
      namespace Request {}
    }
  }
  namespace Messages {
    interface SampleData {
      size?: number;
      id?: string;
      content?: string;
      status?: Enums.SampleStatus;
      createdAt?: Externals.google.protobuf.Timestamp;
      updatedAt?: Externals.google.protobuf.Timestamp[];
      extra?: 100;
      disabledAt?: Externals.google.protobuf.Timestamp;
      deletedAt?: Externals.google.protobuf.Timestamp;
      statusMap?: Record<string, Enums.SampleStatus>;
    }
    namespace SampleData {
      type sampleOneof =
        | { disabledAt: Externals.google.protobuf.Timestamp; deletedAt?: never }
        | { disabledAt?: never; deletedAt: Externals.google.protobuf.Timestamp }
        | { disabledAt?: never; deletedAt?: never };
    }
  }
  namespace Enums {
    type SampleStatus =
      | SampleStatus.UNKNOWN
      | SampleStatus.ACTIVE
      | SampleStatus.INACTIVE;
    namespace SampleStatus {
      type UNKNOWN = 0;
      namespace UNKNOWN {
        const str: "UNKNOWN";
        const num: 0;
      }
      type ACTIVE = 1;
      namespace ACTIVE {
        const str: "ACTIVE";
        const num: 1;
      }
      type INACTIVE = 2;
      namespace INACTIVE {
        const str: "INACTIVE";
        const num: 2;
      }
    }
  }
  namespace Externals {
    namespace google {
      namespace protobuf {
        interface Timestamp {
          seconds?: number;
          nanos?: number;
        }
        namespace Timestamp {}
      }
    }
  }
}

`;

exports[`ProtoToType other formats: strictOneof: hybrid 1`] = `
declare namespace Sample {
  namespace SampleService {
    namespace GetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/get";
      interface Response {
        data?: Messages.SampleData;
      }
      namespace Response {}
      interface Request {
        id?: string;
      }
      namespace Request {}
    }
    namespace SetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/set";
      interface Response {
        status?: Enums.SampleStatus;
      }
      namespace Response {}
      interface Request {
        id?: string;
        data?: Messages.SampleData;
      }
      namespace Request {}
    }
  }
  namespace Messages {
    interface SampleData {
      size?: number;
      id?: string;
      content?: string;
      status?: Enums.SampleStatus;
      createdAt?: Externals.google.protobuf.Timestamp;
      updatedAt?: Externals.google.protobuf.Timestamp[];
      extra?: 100;
      statusMap?: Record<string, Enums.SampleStatus>;
    }
    namespace SampleData {
      type _strict = SampleData & Messages.SampleData.sampleOneof;
      type sampleOneof =
        | { disabledAt: Externals.google.protobuf.Timestamp; deletedAt?: never }
        | { disabledAt?: never; deletedAt: Externals.google.protobuf.Timestamp }
        | { disabledAt?: never; deletedAt?: never };
    }
  }
  namespace Enums {
    type SampleStatus =
      | SampleStatus.UNKNOWN
      | SampleStatus.ACTIVE
      | SampleStatus.INACTIVE;
    namespace SampleStatus {
      type UNKNOWN = "UNKNOWN";
      namespace UNKNOWN {
        const str: "UNKNOWN";
        const num: 0;
      }
      type ACTIVE = "ACTIVE";
      namespace ACTIVE {
        const str: "ACTIVE";
        const num: 1;
      }
      type INACTIVE = "INACTIVE";
      namespace INACTIVE {
        const str: "INACTIVE";
        const num: 2;
      }
    }
  }
  namespace Externals {
    namespace google {
      namespace protobuf {
        interface Timestamp {
          seconds?: number;
          nanos?: number;
        }
        namespace Timestamp {}
      }
    }
  }
}

`;

exports[`ProtoToType other formats: strictOneof: true 1`] = `
declare namespace Sample {
  namespace SampleService {
    namespace GetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/get";
      interface Response {
        data?: Messages.SampleData;
      }
      namespace Response {}
      interface Request {
        id?: string;
      }
      namespace Request {}
    }
    namespace SetSampleData {
      const Endpoint: "(google.api.http).post";
      const Url: "/sample/data/set";
      interface Response {
        status?: Enums.SampleStatus;
      }
      namespace Response {}
      interface Request {
        id?: string;
        data?: Messages.SampleData;
      }
      namespace Request {}
    }
  }
  namespace Messages {
    type SampleData = SampleData._base & Messages.SampleData.sampleOneof;
    namespace SampleData {
      interface _base {
        size?: number;
        id?: string;
        content?: string;
        status?: Enums.SampleStatus;
        createdAt?: Externals.google.protobuf.Timestamp;
        updatedAt?: Externals.google.protobuf.Timestamp[];
        extra?: 100;
        statusMap?: Record<string, Enums.SampleStatus>;
      }
      type sampleOneof =
        | { disabledAt: Externals.google.protobuf.Timestamp; deletedAt?: never }
        | { disabledAt?: never; deletedAt: Externals.google.protobuf.Timestamp }
        | { disabledAt?: never; deletedAt?: never };
    }
  }
  namespace Enums {
    type SampleStatus =
      | SampleStatus.UNKNOWN
      | SampleStatus.ACTIVE
      | SampleStatus.INACTIVE;
    namespace SampleStatus {
      type UNKNOWN = "UNKNOWN";
      namespace UNKNOWN {
        const str: "UNKNOWN";
        const num: 0;
      }
      type ACTIVE = "ACTIVE";
      namespace ACTIVE {
        const str: "ACTIVE";
        const num: 1;
      }
      type INACTIVE = "INACTIVE";
      namespace INACTIVE {
        const str: "INACTIVE";
        const num: 2;
      }
    }
  }
  namespace Externals {
    namespace google {
      namespace protobuf {
        interface Timestamp {
          seconds?: number;
          nanos?: number;
        }
        namespace Timestamp {}
      }
    }
  }
}

`;

exports[`ProtoToType warns if entrypoint wrongs 1`] = `
declare namespace Sample {}

`;
